### ISSUE TITLE
refactor(c23): standardize length-first array APIs and modern pointer contracts

### ISSUE THREAD

**Issue by geisten** [geisten/geistlib#328]
Severity: LOW (maintainability and optimizer contracts)

## Problem

The codebase mixes pointer-first and length-first APIs, legacy null/constant idioms, and inconsistent aliasing contracts. Example public API: `include/geist_arch.h:198-202` takes PCM/output pointers before their lengths/capacities.

For C23 VLA parameter contracts, lengths must precede the arrays they size:

```c
void process(size_t len, int arr[static len]);
```

Cursor bounds are a separate safety concern: `&p[len]` and `p + len` are equivalent and neither is valid for checking an untrusted length. Check `len <= (size_t)(end - p)` first.

## Proposal

- Standardize new/internal array APIs on length/capacity before arrays.
- Add `static`, `const`, and `restrict` only where the full call graph proves their contracts.
- Prefer C23 `bool`, `true`, `false`, `nullptr`, `nullptr_t`, `constexpr`, `typeof`, `auto`, and useful attributes such as `[[nodiscard]]`.
- Replace constant-like macros only where C23 constant expressions preserve required compile-time behavior/ABI.
- Migrate public APIs only with an explicit compatibility/versioning plan.

## Acceptance criteria

- Document the project-wide C23 API style in the contributor guide.
- Migrate internal APIs in bounded batches with tests and benchmark gates.
- Public signature changes include release/ABI migration notes.
- Generated assembly or focused benchmarks show no regression in hot kernels.

## Mandatory C23 implementation rules

- In every modified array API, place all length/capacity parameters before the arrays they describe; use VLA contracts such as `T data[static len]` where the non-null/minimum-size precondition is real.
- Use C23 `bool`, `true`, `false`, and `nullptr`/`nullptr_t` instead of integer booleans or `NULL`. Replace constant macros with typed `constexpr` objects when compile-time and ABI requirements permit.
- Use `[[nodiscard]]` for returned status/size values that must be checked. Add `restrict` only when non-aliasing is proven for every caller. Use `typeof`/`auto` only when they improve clarity without hiding ownership or changing code generation.
- Perform checked `size_t` arithmetic before allocation, indexing, alignment, or pointer advancement. For cursors, validate against `(size_t)(end - p)` before advancing; do not use `p + len` or `&p[len]` as the bounds check.
- Preserve public ABI unless the ticket explicitly includes a versioned compatibility migration.

## Mandatory performance and quality gates

- Migrate APIs in small, independently benchmarkable batches. Establish a parent baseline and rerun the affected microbenchmark plus the standard macOS prefill/decode sweep after every performance-relevant batch.
- Inspect optimized assembly for modified hot functions and record before/after code size and key instruction changes.
- Run the full test and quality suites after each batch. Pure signature/style migrations require bit-identical outputs; do not widen tolerances.
- Any repeatable throughput, latency, memory, binary-size, or quality regression blocks that migration batch.

**Comment by geisten** [comment]
**Batch 1 landed** in 944ecf6 (PR #349), released in **v0.10.3**. Leaving this open — the ticket asks for bounded batches, and 1 of ~6 is done.

### Done

**Style documented** (first acceptance criterion). `CONTRIBUTING.md` gains a *C API style* section: lengths before the arrays they size and why (`T arr[static len]` is only expressible once the length is in scope), where `[static len]` would be a lie and must not be used, `restrict` only where non-aliasing is *proven* rather than expected, subtraction-based cursor bounds, and the migration process itself.

**Batch 1 — the 13 `dequant_*_row` codecs**, 76 call sites across 26 files:

```c
-void dequant_q4_K_row(const void *blocks, float *out, size_t n_elems);
+void dequant_q4_K_row(size_t n_elems, const void *blocks, float out[static n_elems]);
```

The dispatching `dequant_row_fn` typedef moved with them and **stopped being a lie**: it declared `const uint8_t *blocks` while every function takes `const void *`, and nine table entries cast to it. A call through an incompatible function pointer is undefined however compatible the representations look. Making it exact turned two indirect call sites the textual rewrite could not see into compile errors rather than silently reordered arguments.

Also: 81 live-code `NULL` → `nullptr` across 7 files (the 25 that remain are prose in comments, and were deliberately left).

**Verified as a pure reorder**, per the process the same PR documents:

| gate | result |
|---|---|
| opcode sequence, all 11 dequant TUs | **identical** — only argument registers permute (`x2→x0` length, `x0→x1` blocks, `x1→x2` out) |
| `__text` size, all 11 TUs | **byte-identical** |
| `libgeist.a` | 1,362,776 → 1,362,688 B (−88, retired casts) |
| `bench_dequant` best-of-5 | FFN Q4_K 3.56 → 3.36 ms, Q3_K 1.07 → 1.01; Attn 0.93 → 0.85, 0.26 → 0.25 — **noise, not a speedup**; the instructions are the same ones |
| sweep vs parent, back to back | totals 1503.3/4153.3/7825.9 → 1504.5/4165.9/7828.8 ms @128/512/1024 |
| suites | unit 47 / int 56 / py / e2e 8, 0 failures; ASan+UBSan clean; CI 20/20 |

### Remaining — 148 declarations across 28 files

Re-surveyed on `main` at 944ecf6:

| count | file | suggested batch |
|---|---|---|
| 64 | `src/quant/quant.h` | the `linear_*` GEMV/GEMM kernels — **hottest code in the engine**, implementations across NEON/scalar/x86. Its own batch, its own gate. |
| 13 | `src/backends/common/gemma4_kernels.h` | rmsnorm/rope/attention/activation helpers |
| 8 | `src/archs/audio_conformer/encoder_internal.h` | audio encoder internals |
| 8 | `src/archs/transformer/forward/internal.h` | forward-pass helpers |
| 7 | `src/archs/transformer/arch_state.h` | arch state |
| 6 | `include/geist_util.h` | **public** — needs the ABI/versioning plan the ticket calls for, including `geist_arch.h`'s `encode_pcm`, the example in the issue body |
| 4+4+3+… | audio/vision/metal/vulkan internals | small tail |

Suggested order: `gemma4_kernels.h` next (bounded, benchmarkable via the existing kernel benches), then the arch/audio internals, then `quant.h`'s `linear_*` family on its own, and the public headers last behind a versioned migration.

A note for whoever takes batch 2: the rewrite is mechanical but **not idempotent** — running it twice rotates the arguments again. Revert and redo rather than patching a partial pass, and make the dispatching typedef exact *first* so the compiler finds the indirect call sites for you.

**Comment by geisten** [comment]
**Batch 2 landed** in 004fa06 (PR #351), released in **v0.10.4**. Still open — 2 of ~4 batches done, and the remainder is deliberately staged.

### Done in batch 2

**`AGENT.md` — the first acceptance criterion, properly.** Twenty comments across sixteen files already deferred to "AGENT.md" (the `heap.h` allocation rule, the hot-path rule, "no silent truncation", "explicit validation over assertions") and **the file did not exist**. It does now, leading with parameter order, and it is discoverable to tooling: `CLAUDE.md` points at it, `AGENTS.md` symlinks it, CONTRIBUTING and README cross-link it.

**13 kernels + 10 public vtable members, ~340 call sites.** `gemma4_kernels.h` (rmsnorm, rope, both attention variants, the elementwise family, `linear_fp32`, the bf16 helpers) and every public vtable member that took an array before its length — including `encode_pcm`, this issue's opening example.

**Public vtable members with array-before-length: 0 remaining.** All three headers (`geist_arch.h`, `geist_backend.h`, `geist_weight.h`) are `EXPERIMENTAL` and absent from `docs/API_CONTRACT.md`, so no versioned migration was needed. Out-of-tree implementors get a compile error, not silent breakage — verified.

### What the batch taught, now written into AGENT.md

Batch 2 was not mechanical, and three of the lessons cost a CI round-trip each. All are recorded in the `[static len]` table so the next batch does not repeat them:

| trap | why |
|---|---|
| parameter the callee **null-checks** | `[static]` says non-null, the check says maybe — gcc: `-Wnonnull-compare`. The encoder vtable entry points defensively accept null, so they keep plain pointers. |
| bound is a **product over runtime dims** that may be 0 | `float y[static m * n_out]` breaks the moment gcc cannot prove `m > 0`: *"accessing 4 bytes in a region of size 0"* at every call site. `linear_fp32`/`rmsnorm_fp32` document extents in prose instead. |
| nullable parameter | `linear_fp32`'s `bias` (84 call sites) and `rmsnorm_fp32`'s `weight` (12) are passed `nullptr`. |

Also corrected in AGENT.md: the pre-push gcc check must be **`-O3 -c`, never `-fsyntax-only`** — `-Wstringop-overflow`, `-Wstringop-overread` and `-Wnonnull-compare` all come out of the optimizer, so a syntax-only pass reports none of them and CI finds them for you.

### Measured

`[static n]` was isolated from the reorder by compiling one real kernel TU three ways:

| | instructions | bytes |
|---|---|---|
| old order, plain pointers | 2824 | 11591 |
| new order, plain pointers | 2819 | 11571 |
| new order **with** `[static n]` | 2819 | 11571 |

So **`[static n]` changes no code at all**; the reorder itself drops 15 register moves. Net across all touched objects: **+76 bytes** on a 1.36 MB archive. Two objects grew and both were explained rather than waved through — `weight_resolve.o` +24 instructions is *entirely* `mov` (vector count flat at 15, spills flat at 345); `transformer_ops.o` keeps its 172 vector instructions and actually spills *less* (stp/ldp 430→423).

Benchmarked **interleaved A/B** because the host was loaded (load avg 10.9 on 10 cores) and a straight before/after would have read as a 29% regression that was pure contention: three alternating rounds at seq=512 gave −9.1%, −7.7%, −0.7% versus parent — never slower.

Two fixes fell out along the way: `cpu_neon/kernel_catalog.h` re-spelled the kernel signature inline instead of naming the typedef and had silently drifted (it uses `geist_kernel_linear_m1_fn`/`_mN_fn` now and cannot drift again), and gcc caught **113 unchecked `malloc` call sites** in four logits/layer tests feeding possibly-null buffers into the newly-contracted kernels — `tests/test_helpers.h` gained `xmalloc`.

### Remaining — 80 declarations

Re-surveyed on `main` at 004fa06 (was 105 before batch 2):

| count | file | note |
|---|---|---|
| **55** | `src/quant/quant.h` — the `linear_*` GEMV/GEMM family | the big one; see below |
| 3 | `src/quant/quant.h` — `quantize_*` | |
| 4 | `src/archs/audio_conformer/audio_encoder.h` | |
| 4 | `src/archs/transformer/forward/internal.h` | |
| 3 | `src/archs/audio_conformer/audio_kernels.h` | |
| 2 each | vision encoder/kernels, `int4_kv.h`, `cpu_x86/kernel_i2s.h`, … | small tail |

**Batch 3 (suggested):** the ~25 outside `quant.h` — audio, vision, KV, forward internals. Bounded, independently benchmarkable, no ABI surface.

**The `linear_*` family stays deliberately deferred.** 55 declarations, ~390 direct call sites, implementations across NEON/scalar/x86, and it is the hottest code in the engine. The measured payoff for the whole exercise is documentation and diagnostics — nothing in codegen — so churning the inner loop for it is a poor trade unless someone wants it on its own terms. Happy to do it if that is the call; it should be its own PR with its own gate either way.

**Comment by geisten** [comment]
**Batch 3 landed** in 1e785ee (PR #354), released in **v0.10.5**.

### Done

20 helpers across the arch and backend-common layers, **122 call sites**: the `attention_{int4,int8,kivi}_via_buffers` family (lengths had been scattered through up to twenty parameters), `kivi_drain_one_layer`, the audio encoder entry points, the vision tower runners, and the small elementwise / FWHT / int4-KV helpers. Internal signatures only — no public header, no ABI change.

**The batch-2 traps were checked before pushing, not discovered in CI.** Every candidate was tested against all three rules from AGENT.md — single dimension, no defensive null-check in the callee, no `nullptr` call site. Exactly **five parameters passed all three** and got `[static n]`; everything else keeps a plain pointer deliberately, mostly because the extent is a product over runtime dimensions (`frames × n_mels`, `tokens × heads × head_dim`) that batch 2 established gcc will not accept. `gcc-15 -O3` was clean **before** the push — which is the entire point of the correction AGENT.md picked up last round, since the same check at `-fsyntax-only` reports none of these.

**Codegen** — vector instruction count identical in all eight touched objects, spills net −3, +24 bytes total:

| object | text | vector | spills |
|---|---|---|---|
| `audio_kernels.o` | 5924 → 5924 | 98 → 98 | 197 → 197 |
| `encoder_forward.o` | 20727 → 20723 | 368 → 368 | 566 → 566 |
| `encoder_stream.o` | 14800 → 14800 | 292 → 292 | 228 → 228 |
| `vision_encoder.o` | 8909 → 8925 | 11 → 11 | 162 → 162 |
| `vision_kernels.o` | 6763 → 6759 | 30 → 30 | 200 → 201 |
| `attention.o` | 11391 → 11399 | 108 → 108 | 259 → **258** |
| `kv_store.o` | 6634 → 6642 | 54 → 54 | 104 → **101** |
| `fwht.o` | 508 → 508 | 6 → 6 | 8 → 8 |

The wall-clock benchmark was **not** offered as evidence: the host sat at load 8.6 and interleaved A/B gave −0.7%, +2.1%, −11.2% at seq=512 — a spread far wider than any effect this change could produce.

### One function I missed

`vision_encoder_run_tower` should have been in this batch and was not:

```c
bool vision_encoder_run_tower(const struct VisionEncoder *, const float *patches_in,
                              const int32_t *positions, size_t n_patches, float *hidden_out);
```

My extraction pass matched a comment line mentioning it rather than the declaration, and I concluded from that it was not a real declaration. It is. The fix is `n_patches` to the front, 5 sites total (declaration, definition, two internal calls, one test). Not folded into batch 3 after the fact; it belongs in the next batch or a one-line follow-up.

### Remaining — 61 declarations

Re-surveyed on `main` at 1e785ee (was 80 before batch 3):

| count | where | note |
|---|---|---|
| **58** | `src/quant/quant.h` | 55 `linear_*` GEMV/GEMM + 3 `quantize_*` |
| 2 | `cpu_x86/kernel_i2s.h` | **false positives** — `i2s_x4_gemv_pair_m1` / `i2s_t5_gemv_pair_m1` already place `n_out0` before the `y0` it sizes; what precedes it is `x4_0`, which `n_out0` does not size. Left deliberately. |
| 1 | `vision_encoder.h` | the miss above |

So the genuine remainder is **59**, and 58 of those are one file.

For scale: the tree now carries **345 `[static len]` contracts**, up from 268 when this issue was opened.

### The `linear_*` family — the open decision

This is what is actually left, and it has not been deferred out of caution alone. The numbers from three batches of measuring:

- ~390 direct call sites, implementations across NEON / scalar / x86, the hottest code in the engine.
- `[static n]` has been shown here to change **no code at all** (2819 instructions and 11571 bytes with and without, same parameter order).
- Most of the family's extents are products over `m`, which gcc rejects as a contract anyway — so much of the family would end up with plain pointers regardless, leaving only the reorder.

That is a large, risky diff in the inner loop for a payoff that is documentation and uniformity. Worth doing if the goal is that every signature in the tree reads the same way — but it should be an explicit call, not a tail-end of a batch. Happy to take it on with its own PR and its own disassembly gate if that is the decision.

**Comment by geisten** [comment]
**Batch 4 is ready in PR #377.** This closes the genuine small tail from batch 3: the three `quantize_*` declarations in `src/quant/quant.h` plus the missed `vision_encoder_run_tower`, with every caller updated.

Contracts are deliberately truthful: `quantize_x_int8_sym` and `quantize_x_for_q4k` now use length-first `[static n]`; `quantize_x_for_q4k_blocks` and the vision entry point retain plain pointers because their bodies defensively accept/check null. No public header or ABI changes.

### Gates

- release unit: 49 total, 29 pass / 20 skip / 0 fail
- release integration: 56 total, 25 pass / 31 fixture skip / 0 fail
- model fixtures: Q4_K, Q6_K, PTQTP release exactness, and all 16 vision layers pass
- ASan+UBSan: unit plus Q4_K, Q6_K, and 16-layer vision pass
- gcc 15 `-O3 -Werror -Wshadow -Wundef`: all changed production TUs clean
- format 343/343; diff check clean
- optimized `common.o` 4,416 → 4,416 B and `q4_K.o` 45,744 → 45,744 B; symbol offsets, instruction counts (172 and 123), and opcode sequences are identical. Only argument registers permute.
- paired M1 Max parent/patch benchmark, Gemma 4 E2B Q4_K_M, 8 threads, 64 warm-up, mean-of-5, 128 decode tokens:
  - seq128 total 69.431 → 70.317 t/s (+1.28%)
  - seq512 total 93.210 → 93.325 t/s (+0.12%)
  - individual prefill deltas +0.14%/+0.73%; decode +1.57%/-0.40%
  - no repeatable regression

The ASan `-O1` PTQTP fixture exposes a pre-existing exact GEMM/GEMV mismatch that reproduces unchanged on the parent; release remains 0 mismatches. It is recorded with acceptance criteria in #376 rather than hidden or absorbed by a tolerance change.

`ROADMAP.md` now reflects the actual closed review backlog and current open work. The remaining #328 scope is exactly the 55-function hot `linear_*` family, still isolated for its own PR and assembly/performance gate.
